File: Strings/firstUniqChar.cpp

// alpha is used to keep track of the lowest index each letter occurs
// If the same letter is encountered twice mark it's lowest index as 
// one that is beyond the array (i.e. can't possible be the minIndex)

// Iterate through alpha, comparing the indices of all the unique chars
// and return the minimum

// Time complexity      O(n)
// Space complexity     O(1)
class Solution {
public:
    int firstUniqChar(string s) {
        int n = s.size();
        
        int alpha[26] = {[0 ... 25] = n};
    
        
        for (int i = 0; i < n; ++i){
            int index = s[i] - 97;
            if (alpha[index] == n){
                alpha[index] = i;
            }
            else if (alpha[index] < n){
                alpha[index] = n + 1;
            }
        }
        
        int minIndex = n;
        
        for (int i = 0; i < 26; ++i){
            minIndex = min(alpha[i], minIndex);
        }
        
        return minIndex != n ? minIndex : -1;
    }
};

//Version 2
// repeat the above but rather than iterating through s once and alpha once
// iterate through s twice, on the second run checking for the first char whose count is 1

File: Strings/firstUniqChar.py
# Version 1 - standard hashmap
# Time complexity       O(n)
# Space complexity      O(1)
class Solution:
    def firstUniqChar(self, s: str) -> int:
        cnt = collections.defaultdict(lambda:[])
        
        n = len(s)
        for i in range(n):
            cnt[s[i]].append(i)
        
        minIndex = n
        for idxList in cnt.values():
            if len(idxList) == 1:
                minIndex = min(idxList[0], minIndex)
        
        if minIndex < n:
            return minIndex
        else:
            return -1